libgalois/include/katana/PropertyIndex.h

```diff
@@ -0,0 +1,88 @@
+#ifndef KATANA_LIBGALOIS_KATANA_PROPERTYINDEX_H_
+#define KATANA_LIBGALOIS_KATANA_PROPERTYINDEX_H_
+
+#include <bitset>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <arrow/api.h>
+#include <arrow/chunked_array.h>
+#include <arrow/type_traits.h>
+#include <boost/iterator/iterator_categories.hpp>
+#include <boost/iterator/iterator_facade.hpp>
+
+#include "katana/config.h"
+#include "katana/PropertyGraph.h"
+#include "katana/Result.h"
+
+namespace katana {
+
+// PropertyIndex provides an interface similar to an ordered container
+// over a single property.
+class KATANA_EXPORT PropertyIndex {
+public:
+  // PropertyIndex::iterator returns a sequence of node or edge ids.
+  using node_iterator = std::set<GraphTopology::Node>::iterator;
+  using edge_iterator = std::set<GraphTopology::Edge>::iterator;
+
+  // Create an index from a column of pg.
+  static katana::Result<std::unique_ptr<PropertyIndex>>
+    Make(PropertyGraph* pg, std::string column);
+
+  std::string column() { return column_; }
+
+  virtual node_iterator node_begin() = 0;
+  virtual node_iterator node_end() = 0;
+
+  virtual edge_iterator edge_begin() = 0;
+  virtual edge_iterator edge_end() = 0;
+
+private:
+  std::string column_;
+};
+
+// BasicPropertyIndex provides a PropertyIndex for anything that can be used
+// in a default std::set.
+template<typename key_type>
+class KATANA_EXPORT BasicPropertyIndex : public PropertyIndex {
```

Comment:
I think it would make sense to decouple the `node` and `edge` properties. A `BasicPropertyIndex` should have just one column of values over which the index is maintained. Unless i am mistaken, we do not expect the same property to exist on both the node and edges.

Reply:
I wrote this iteration this way to avoid templating altogether. Based on some further conversation with Daniel, I've returned to a model where node/edge becomes part of the templated type of the index. So you will have either PropertyIndex<Node> or PropertyIndex<Edge>

Reply:
Haha, doesn't work well if you try to write template arguments in html.

`PropertyIndex<Node> or PropertyIndex<Edge>`